### DISCUSSION TITLE
Found zeroed-out block fix

### DISCUSSION THREAD

**Discussion by franz101** [jivanpal/apfs-tools#28] [Q&A]
Hey Jivan,
First of all hats of to the amazing work you have done.

I seem to have the same issue as you had and one other in the discussions....

![IMG_20210926_001913](https://user-images.githubusercontent.com/18228395/134799232-24863077-3b0a-4c49-b6c9-ea06a9e9c338.jpg)
![IMG_20210926_001852](https://user-images.githubusercontent.com/18228395/134799237-29845017-5254-420a-8c6e-d59a8871edaa.jpg)

 I wasn't able to run recover. Once I have a second hard drive, I will try again.

`error: (oid Ox1cb5ead) cib: found zeroed-out block error: failed to read spaceman cib 0x1cb5ead

Space manager is invalid.

The volume /dev/rdisk2s5 could not be verified completely.`

My question is should I wait for 0.2? Or try to recover? Though I get stuck with the execution order of the commands and physical ids or virtual id.

How can I create the file list.txt for pull.sh?

If I try recover with a file path I know exists it gives me an error. Or is it the wrong /dev/disk?

**Comment by franz101** [comment]
![16326441394168870877538052235855](https://user-images.githubusercontent.com/18228395/134799623-ca8b3ee2-1428-43a5-a7c8-40e22d8f6206.jpg)

**Reply by franz101** [reply]
So listing files in my home.directory does not work....
It works for other users....
But if I try i get 
get_fs_records: Need to descend to node with Virtual OID exdd7fc3, but the volume object map lists no objects with this virtual 010.

**Reply by jivanpal** [reply]
> get_fs_records: Need to descend to node with Virtual OID exdd7fc3, but the volume object map lists no objects with this virtual 010.

Looks like your volume's omap is corrupt. It will need to be repaired in order for v0.1.3's `recover` to work. v0.2 will have better support for past APFS transactions and thus file recovery, but in the mean time, you can try the free version of [Disk Drill](https://www.cleverfiles.com/) to see if your files can be found (which Drat v0.2 or the paid version of Disk Drill should then be able to recover).

> How can I create the file list.txt for pull.sh?

See #13 for some info on how to do this, but since your volume omap is corrupt, you won't be able to do anything useful without fixing that first anyway. v0.2 will remove the need to use `pull.sh`; that functionality will be incorporated into `recover`.

For now, try the following commands to see where the filesystem lookup for `/Users/franz/.aws/config` is failing:

- `drat list /dev/disk0s2 0 /`
- `drat list /dev/disk0s2 0 /Users`
- `drat list /dev/disk0s2 0 /Users/franz`
- `drat list /dev/disk0s2 0 /Users/franz/.aws`
- `drat list /dev/disk0s2 0 /Users/franz/.aws/config`

If each command is successful, the contents of the corresponding directory should be listed. If unsuccessful, you should get a `get_fs_records` error as before.

**Reply by franz101** [reply]
Great. Awesome job. So yes home can be listed. But the user can't be listed... First did a byte to byte back up with disk drill. Which results in a dmg.
And now running their deep recovery scan. It takes around 37 hours....

![IMG_20210926_204431](https://user-images.githubusercontent.com/18228395/134820255-70c9f234-d9e0-4849-b15f-bca2509f0e38.jpg)

**Comment by zhanchengqian** [comment]
@jivanpal Hi Jivan, I just want to come by and say hi. I later found a software R-studio (r-tt.com). Its website looks quite sketchy but I tried it and it actually listed 99% of the missing files. @franz101 If you want to try other data recovery software, I would highly recommend R-studio (r-tt.com). This is not an Ad, it's just my personal experience.
